File: Chain/Arcana.swift
//
//  Arcana.swift
//  Chain
//
//  Created by Jitae Kim on 8/25/16.
//  Copyright © 2016 Jitae Kim. All rights reserved.
//

import Firebase

struct Arcana: Equatable, Hashable {
    
    var uid: String
    var nameKR: String
    var nickNameKR: String?
    var nameJP: String
    var nickNameJP: String?
    var rarity: String
    var group: String  // Class, 직업
    var tavern: String  // 주점
    var affiliation: String? // 소속
    var cost: String
    var weapon: String

    var kizunaName: String
    var kizunaCost: String
    var kizunaDesc: String
    
    var skillCount : String
    
    var skillName1 : String
    var skillMana1 : String
    var skillDesc1 : String
    
    var skillName2 : String?
    var skillMana2 : String?
    var skillDesc2 : String?
    
    var skillName3 : String?
    var skillMana3 : String?
    var skillDesc3 : String?
    
    
    var abilityName1: String?
    var abilityDesc1: String?
    
    var abilityName2: String?
    var abilityDesc2: String?
    
    var numberOfViews: Int
    var dateAdded: String?
    var chainStory: String?
    var chainStone: String?
    
    var hashValue: Int {
        return uid.hashValue
    }
    
    

    
    
    init?(u: String, nKR: String, nnKR: String, nJP: String, nnJP: String, r: String, g: String, t: String, a: String, c: String, w: String, kN: String, kC: String, kD: String, sC: String, sN1: String, sM1: String, sD1: String, sN2: String, sM2: String, sD2: String, sN3: String, sM3: String, sD3: String, aN1: String, aD1: String, aN2: String, aD2: String, v: Int) {
        
        uid = u
        nameKR = nKR
        nickNameKR = nnKR
        nameJP = nJP
        nickNameJP = nnJP
        rarity = r
        group = g
        tavern = t
        affiliation = a
        cost = c
        weapon = w
        
        kizunaName  = kN
        kizunaCost = kC
        kizunaDesc = kD
        
        skillCount = sC
        
        skillName1 = sN1
        skillMana1 = sM1
        skillDesc1 = sD1
        
        skillName2 = sN2
        skillMana2 = sM2
        skillDesc2 = sD2
        
        skillName3 = sN3
        skillMana3 = sM3
        skillDesc3 = sD3
        
        abilityName1 = aN1
        abilityDesc1 = aD1
        
        abilityName2 = aN2
        abilityDesc2 = aD2
        
        numberOfViews = v
        
        
    }
    

 
    init?(snapshot: FIRDataSnapshot) {
        
//        if let a = snapshot.value as? NSDictionary {
//            let b = a["uid"] as? String
//        }
        guard let u = (snapshot.value as? NSDictionary)?["uid"] as? String, let nKR = (snapshot.value as? NSDictionary)?["nameKR"] as? String, let nJP = (snapshot.value as? NSDictionary)?["nameJP"] as? String, let r = (snapshot.value as? NSDictionary)?["rarity"] as? String, let g = (snapshot.value as? NSDictionary)?["class"] as? String, let t = (snapshot.value as? NSDictionary)?["tavern"] as? String, let a = (snapshot.value as? NSDictionary)?["affiliation"] as? String, let c = (snapshot.value as? NSDictionary)?["cost"] as? String, let w = (snapshot.value as? NSDictionary)?["weapon"] as? String, let kN = (snapshot.value as? NSDictionary)?["kizunaName"] as? String, let kC = (snapshot.value as? NSDictionary)?["kizunaCost"] as? String, let kD = (snapshot.value as? NSDictionary)?["kizunaDesc"] as? String, let sC = (snapshot.value as? NSDictionary)?["skillCount"] as? String, let sN1 = (snapshot.value as? NSDictionary)?["skillName1"] as? String, let sM1 = (snapshot.value as? NSDictionary)?["skillMana1"] as? String, let sD1 = (snapshot.value as? NSDictionary)?["skillDesc1"] as? String, let v = (snapshot.value as? NSDictionary)?["numberOfViews"] as? Int else {
                print("COULD NOT GET SNAPSHOT OF 1 SKILL ARCANA")
                return nil
            }
        
            uid = u
            nameKR = nKR
            nameJP = nJP
            rarity = r
            group = g
            tavern = t
            affiliation = a
            cost = c
            weapon = w
            
            kizunaName  = kN
            kizunaCost = kC
            kizunaDesc = kD
            
            skillCount = sC
            
            skillName1 = sN1
            skillMana1 = sM1
            skillDesc1 = sD1
            
            numberOfViews = v
        
        if let nnKR = (snapshot.value as? NSDictionary)?["nickNameKR"] as? String, let nnJP = (snapshot.value as? NSDictionary)?["nickNameJP"] as? String {
            nickNameKR = nnKR
            nickNameJP = nnJP
        }
        
        if let aN1 = (snapshot.value as? NSDictionary)?["abilityName1"] as? String, let aD1 = (snapshot.value as? NSDictionary)?["abilityDesc1"] as? String {
            abilityName1 = aN1
            abilityDesc1 = aD1
        }
        
        if let aN2 = (snapshot.value as? NSDictionary)?["abilityName2"] as? String, let aD2 = (snapshot.value as? NSDictionary)?["abilityDesc2"] as? String {
            abilityName2 = aN2
            abilityDesc2 = aD2
        }
        if let sN2 = (snapshot.value as? NSDictionary)?["skillName2"] as? String, let sM2 = (snapshot.value as? NSDictionary)?["skillMana2"] as? String, let sD2 = (snapshot.value as? NSDictionary)?["skillDesc2"] as? String {
            skillName2 = sN2
            skillMana2 = sM2
            skillDesc2 = sD2
        }
        
        if let sN3 = (snapshot.value as? NSDictionary)?["skillName3"] as? String, let sM3 = (snapshot.value as? NSDictionary)?["skillMana3"] as? String, let sD3 = (snapshot.value as? NSDictionary)?["skillDesc3"] as? String {
            skillName3 = sN3
            skillMana3 = sM3
            skillDesc3 = sD3
        }
        
        if let d = (snapshot.value as? NSDictionary)?["dateAdded"] as? String {
            dateAdded = d
        }
        
        if let cStory = (snapshot.value as? NSDictionary)?["chainStory"] as? String {
            chainStory = cStory
        }
        
        if let cStone = (snapshot.value as? NSDictionary)?["chainStone"] as? String {
            chainStone = cStone
        }

        
  
    }
    

    
}

func ==(lhs: Arcana, rhs: Arcana) -> Bool {
    return lhs.uid == rhs.uid
}
struct ArcanaDictionary {
    var arcanaDictionary: [Arcana]?
}

File: functions/public/js/arcana.js
/**
 * Copyright 2015 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
'use strict';

var arcanaID;

window.onload = function() {
  window.ChainWiki = new ChainWiki();
};

// Initializes ChainWiki.
function ChainWiki() {
  // Shortcuts to DOM Elements.
  this.arcanaList = document.getElementById('arcana');
  this.initFirebase();
  arcanaID = getParameterByName('arcana');
  this.loadArcana();
  if (location.hostname !== "localhost" && location.hostname == "172.91.86.210") {
    incrementViewCount();
  }
}

// Sets up shortcuts to Firebase features and initiate firebase auth.
ChainWiki.prototype.initFirebase = function() {
  // Shortcuts to Firebase SDK features.
  this.auth = firebase.auth();
  this.database = firebase.database();
  this.storageRef = firebase.storage().ref();
};

function incrementViewCount() {
  firebase.database().ref(`arcana/${arcanaID}/numberOfViews`).transaction(function(count) {
    if (count) {
      console.log('count was', count);
      count++;
    }
    return count;
  });

}

ChainWiki.prototype.loadArcana = function() {
  console.log(arcanaID);
  
  this.arcanaRef = this.database.ref(`arcana/${arcanaID}`);
  // Make sure we remove all previous listeners.
  this.arcanaRef.off();

  var setupImage = function(arcanaID) {
    this.loadImage(arcanaID);
  }.bind(this);

  var setupBaseInfo = function(val) {
    this.setupBaseInfo(val);
  }.bind(this);

  this.arcanaRef.once('value').then(function(snapshot) {

    const val = snapshot.val();
    const arcanaID = snapshot.key;
    document.title = val.nameKR;
    setupImage(arcanaID);
    setupBaseInfo(val);
  });
};

ChainWiki.prototype.setupBaseInfo = function(val) {
  
  document.getElementById('nameKR').innerHTML = val.nameKR;
  document.getElementById('rarity').innerHTML = val.rarity;
  document.getElementById('group').innerHTML = val.class;
  document.getElementById('affiliation').innerHTML = val.affiliation;
  document.getElementById('cost').innerHTML = val.cost;
  document.getElementById('weapon').innerHTML = val.weapon;
  document.getElementById('tavern').innerHTML = val.tavern;

  document.getElementById('skillName1').innerHTML = val.skillName1;
  document.getElementById('skillMana1').innerHTML = val.skillMana1;
  document.getElementById('skillDesc1').innerHTML = val.skillDesc1;

  if (val.skillDesc2) {
    addSkill(val.skillName2, val.skillMana2, val.skillDesc2, 2);
  }

  if (val.skillDesc3) {
    addSkill(val.skillName3, val.skillMana3, val.skillDesc3, 3);
  }

  document.getElementById('abilityName1').innerHTML = val.abilityName1;
  document.getElementById('abilityDesc1').innerHTML = val.abilityDesc1;

  if (val.abilityName2) {
    addAbility(val.abilityName2, val.abilityDesc2, 2);
  }

  if (val.partyAbility) {
    addAbility(val.partyAbility, val.partyAbility, 'partyAbility');
  }

  document.getElementById('kizunaName').innerHTML = val.kizunaName;
  document.getElementById('kizunaCost').innerHTML = val.kizunaCost;
  document.getElementById('kizunaDesc').innerHTML = val.kizunaDesc;

};

function insertAfter(referenceNode, newNode) {
  referenceNode.parentNode.insertBefore(newNode, referenceNode.nextSibling);
}

function addSkill(skillName, skillMana, skillDesc, count) {

  // create the one row table.
    
    const table = document.createElement('table');
    table.setAttribute('class', 'arcanaSkillTable');

    const row = table.insertRow(0);

    const headerCell = document.createElement('th');
    headerCell.setAttribute('class', 'headerCell');

    const bodyCell = document.createElement('td');
    bodyCell.setAttribute('class', 'bodyCell');

    row.appendChild(headerCell);
    row.appendChild(bodyCell);

    const skillContainerView = document.createElement('div');
    skillContainerView.setAttribute('class', 'container');
    
    const skillNameCell = document.createElement('div');
    skillNameCell.setAttribute('class', 'skillNameCell');
    skillNameCell.style.cssFloat = 'left';
    skillNameCell.innerHTML = skillName;

    const skillManaCell = document.createElement('div');
    skillManaCell.setAttribute('class', 'manaCell');
    skillManaCell.style.cssFloat = 'right';
    skillManaCell.innerHTML = skillMana;

    bodyCell.appendChild(skillContainerView);
    skillContainerView.appendChild(skillNameCell);
    skillContainerView.appendChild(skillManaCell);

    // setup the skillDesc below the table
    const skillDescDiv = document.createElement('div');
    skillDescDiv.setAttribute('class', 'skillAbilityDescCell');
    skillDescDiv.innerHTML = skillDesc;
    if (count == 2) {
      skillDescDiv.setAttribute('id', 'skillDesc2');
    }

    var previousDiv;
    if (count == 2) {
      console.log('adding skill2');
      headerCell.innerHTML = '스킬 2';
      // skillNameCell.setAttribute('id', 'skillName1');
      
      previousDiv = document.getElementById('skillDesc1');
    }
    else if (count == 3) {
      console.log('adding skill3');
      headerCell.innerHTML = '스킬';
      previousDiv = document.getElementById('skillDesc2');
    }
    insertAfter(previousDiv, table);
    insertAfter(table, skillDescDiv);
}

function addAbility(abilityName, abilityDesc, type) {

    // create the one row table.
    const table = document.createElement('table');
    table.setAttribute('class', 'arcanaSkillTable');

    const row = table.insertRow(0);

    const headerCell = document.createElement('th');
    headerCell.setAttribute('class', 'headerCell');

    const bodyCell = document.createElement('td');
    bodyCell.setAttribute('class', 'bodyCell');

    row.appendChild(headerCell);
    row.appendChild(bodyCell);

    // setup the abilityDesc below the table
    const abilityDescDiv = document.createElement('div');
    abilityDescDiv.setAttribute('class', 'skillAbilityDescCell');
    abilityDescDiv.setAttribute('id', 'abilityDesc2');
    abilityDescDiv.innerHTML = abilityDesc;

    var previousDiv;
    if (type == 2) {
      console.log('adding ability2');
      bodyCell.innerHTML = abilityName;
      headerCell.innerHTML = '어빌 2';
      previousDiv = document.getElementById('abilityDesc1');
    }
    else {
      console.log('adding partyability');
      headerCell.innerHTML = '파티 어빌';
      previousDiv = document.getElementById('abilityDesc2');
    }
    insertAfter(previousDiv, table);
    insertAfter(table, abilityDescDiv);
}
ChainWiki.prototype.loadImage = function(arcanaID) {

  const arcanaImage = document.getElementById('arcanaImage');
  this.storageRef.child(`/image/arcana/${arcanaID}/main.jpg`).getDownloadURL().then(function(url) {
    arcanaImage.innerHTML = `<img src =${url} class="arcanaImageMain"/>`;
    $('img.arcanaImageMain').lazyload({
      // placeholder: 'images/placeholder.png',
      effect : 'fadeIn'
    });
  }).catch(function(error) {
    console.log("Image download error.");
  });
};

function getParameterByName(name, url) {
    if (!url) {
      url = window.location.href;
    }
    name = name.replace(/[\[\]]/g, "\\$&");
    var regex = new RegExp("[?&]" + name + "(=([^&#]*)|&|#|$)"),
        results = regex.exec(url);
    if (!results) return null;
    if (!results[2]) return '';
    return decodeURIComponent(results[2].replace(/\+/g, " "));
}